torch/csrc/utils/python_numbers.h

```diff
@@ -66,28 +77,17 @@ inline bool THPUtils_checkDouble(PyObject* obj) {
 }
 
 inline double THPUtils_unpackDouble(PyObject* obj) {
-  if (PyFloat_Check(obj)) {
-    return PyFloat_AS_DOUBLE(obj);
+  // PyFloat_AsDouble will calls obj.__float__ if obj is not a float object
+  auto value = PyFloat_AsDouble(obj);
+  if (PyErr_Occurred()){
+    throw python_error();
   }
+  // raise warning when converting a long to float with precision loss
   if (PyLong_Check(obj)) {
-    int overflow;
-    long long value = PyLong_AsLongLongAndOverflow(obj, &overflow);
-    if (overflow != 0) {
-      throw std::runtime_error("Overflow when unpacking double");
-    }
     if (value > DOUBLE_INT_MAX || value < -DOUBLE_INT_MAX) {
```

Comment:
How is `value` supposed to be out of range of `double` if it's a double itself?

Reply:
The original overflow message is not needed here.

early version:
`unpackDouble(python_int) --> c_long_long(python_int) --> c_double(c_long_long)`
unpackDouble(int(1e+40)) would raise the **overflow** error because long long  (typically 64 bits) can only represent up to **2^127**.  However the 1e+40 is totally representable by a `double` type.

Using double to represent `int(1e+40)` causes problem of **precision loss** as the absolute value is too large (here `DOUBLE_INT_MAX`). Compared with overflow, I think **precision loss** is somehow acceptable so a *warning* will be raised instead of *error*

this version:
`unpackDouble(python_int) --> python_float(python_int) --> c_double(python_float)`
The type conversions are all handled by cpython, so we can simply check `PyErr` and return `NULL` on error.